src/pybind/mgr/cephadm/services/nvmeof.py

```diff
@@ -47,12 +47,8 @@ def prepare_create(self, daemon_spec: CephadmDaemonDeploySpec) -> CephadmDaemonD
         transport_tcp_options = json.dumps(spec.transport_tcp_options) if spec.transport_tcp_options else None
         name = '{}.{}'.format(utils.name_to_config_section('nvmeof'), nvmeof_gw_id)
         rados_id = name[len('client.'):] if name.startswith('client.') else name
-        addr = host_ip
-        discovery_addr = host_ip
-        if spec.addr and spec.addr != "0.0.0.0":
-            addr = spec.addr
-        if spec.discovery_addr and spec.discovery_addr != "0.0.0.0":
-            discovery_addr = spec.discovery_addr
+        addr = spec.addr or host_ip
+        discovery_addr = spec.discovery_addr or host_ip
```

Comment:
it's important to note you're effectively changing the defaults here. Since these fields default to `0.0.0.0` in the spec, due to the code you're removing, they would actually default to the host ip. This change is saying you want the nvmeof conf to be populated with `0.0.0.0` addresses by default for all nvmeof daemons rather than the IPs of the host they're deployed on. Is that what we're going for here?

Reply:
@adk3798 🙏 missed that.

how about [this](https://github.com/ceph/ceph/pull/59755/commits/0997e4cd51edfe954148bfb8b7f42be9205c237c) modification, it would allow cephadm service to determine default address, i.e. host ip and also alow user to set '0.0.0.0' via spec. would it break anything? wdyt?

Reply:
I think with that additional change this should work as intended. Default to host ip but use the user entered addr / discovery addr when provided (including if the user entered addr is 0.0.0.0)

Reply:
There is still a potential risk for those who are upgrading and already had the default 0.0.0.0 address set due to past defaults. But I'm not sure there's much way around that as it becomes difficult to determine if they just have the past defaults or they actually wanted 0.0.0.0. I'm not sure there's any good way around that though. Just making you aware we will see that on upgrades.

Reply:
@adk3798, Looking at the 7.1 downstream code the default runtime value of the address would be host ip, but the spec exported value will be '0.0.0.0'?  Is it correct? 

On the positive side, looks like after upgrade, spec would match runtime address value. However the upgrade should succeed.

Reply:
yeah, if the previous spec defaulted to `0.0.0.0` and then we remove that default, specs that already had `0.0.0.0` from the past default will retain that as the spec has no way of knowing it was from a past default when loaded in the new code. By `spec would match runtime address value` do you mean that new specs will work fine? The upgrade will probably succeed, the users just might end up with `0.0.0.0` in their addr and discovery addr fields in the nvmeof conf for each nvmeof daemon (unless that causes it to fail to start, in which case the upgrade will actually fail).